src/number_theory/legendre_symbol/gauss_sum.lean

```diff
@@ -0,0 +1,414 @@
+begin
+  have h₀ : gauss_sum χ ψ ≠ 0 :=
+  begin
+    intro hf,
+    simp [hf] at hg,
+    have hh := or.resolve_left hg (is_unit.map χ is_unit_one.neg).ne_zero,
+    rw [← show ((fintype.card R : ℤ) : R') = fintype.card R, by norm_cast] at hh,
+    exact not_is_unit_prime_of_dvd_card p
+            (int.coe_nat_dvd.mp ((char_p.int_cast_eq_zero_iff R' p (fintype.card R)).mp hh)) hp,
+  end,
+  have h₁ := quad_gauss_sum_frob_iter p n hp hχ ψ,
+  have h₂ := nat.odd_iff.mp (odd.pow (or.resolve_left (nat.prime.eq_two_or_odd' fp.1) hp')),
+  rw [← nat.div_add_mod (p ^ n) 2, h₂, pow_succ'] at h₁,
+  rw [← mul_right_cancel₀ h₀ h₁, pow_mul, hg],
+end
+
+/-- When `F` and `F'` are finite fields and `χ : F → F'` is a nontrivial quadratic character,
+then `(χ (-1) * #F)^(#F'/2) = χ (#F')`. -/
+lemma char.card_pow_card {F : Type} [field F] [fintype F] {F' : Type} [field F'] [fintype F']
+  {χ : mul_char F F'} (hχ₁ : is_nontrivial χ) (hχ₂ : is_quadratic χ)
+  (hch₁ : ring_char F' ≠ ring_char F) (hch₂ : ring_char F' ≠ 2) :
+  (χ (-1) * fintype.card F) ^ (fintype.card F' / 2) = χ (fintype.card F') :=
+begin
+  obtain ⟨n, hp, hc⟩ := finite_field.card F (ring_char F),
+  obtain ⟨n', hp', hc'⟩ := finite_field.card F' (ring_char F'),
+  let ψ := primitive_char_finite_field F F' hch₁,
+  let FF' := cyclotomic_field ψ.n F',
+  have hchar := -- `ring_char FF' = ring_char F'`
+    ring_char.eq_iff.mpr ((algebra.char_p_iff F' FF' (ring_char F')).mp (ring_char.char_p F')),
+  haveI : fact (ring_char FF').prime := by { rw ← hchar at hp', exact ⟨hp'⟩ },
+  let χ' := χ.ring_hom_comp (algebra_map F' FF'),
+  have hχ' : ∀ x : F, (algebra_map F' FF') (χ x) = χ' x := λ x, rfl,
+  have hχ₁' := hχ₁.comp (algebra_map F' FF').injective,
+  have hχ₂' := hχ₂.comp (algebra_map F' FF'),
+  have hpu : is_unit (ring_char FF' : F) :=
+  begin
+    rw [hchar],
+    haveI := invertible_of_ring_char_not_dvd
+              (λ hf, hch₁ ((nat.prime_dvd_prime_iff_eq hp hp').mp hf).symm),
+    exact is_unit_of_invertible (ring_char F' : F),
+  end,
+  apply_fun algebra_map F' FF' using (algebra_map F' FF').injective,
+  rw [map_pow, map_mul, hχ', hχ', map_nat_cast, hc', ← hchar,
+      char.card_pow_char_pow hχ₂' ψ.char (ring_char FF') n' hpu (ne_of_eq_of_ne hchar hch₂)
+        (gauss_sum_sq hχ₁' hχ₂' ψ.prim),
+      nat.cast_pow],
+end
+
+end gauss_sum_values
+
+/-!
+### The quadratic character of 2
+
+This section proves the following result.
+
+For every finite field `F` of odd characteristic, we have `2^(#F/2) = χ₈(#F)` in `F`.
+This can be used to show that the quadratic character of `F` takes the value
+`χ₈(#F)` at `2`.
+
+The proof uses the Gauss sum of `χ₈` and a primitive additive character on `ℤ/8ℤ`;
+in this way, the result is reduced to `card_pow_char_pow`.
+-/
+
+section gauss_sum_two
+
+/-- Expand a sum over `ℤ/8ℤ` -/
+private
+lemma sum8 {R : Type*} [add_comm_monoid R] (f : (zmod 8) → R) :
+  ∑ a, f a = f 0 + f 1 + f 2 + f 3 + f 4 + f 5 + f 6 + f 7 :=
```

Comment:
I don't think this should be `private`. We haven't needed it. But since you need it now, I wouldn't be surprised if some one else will also need it in the future. Please add a version for `fin` next to `fin.sum_univ_two`.

Reply:
I assume it would then make sense to also fill the gap between two and eight. I'll add this soon.

Reply:
Done now.